front/lib/resources/content_fragment_resource.ts

```diff
@@ -276,3 +284,102 @@ export async function getContentFragmentText({
 
   return getPrivateUploadBucket().fetchFileContent(filePath);
 }
+
+async function getSignedUrlForRawContentFragment({
+  workspaceId,
+  conversationId,
+  messageId,
+}: {
+  workspaceId: string;
+  conversationId: string;
+  messageId: string;
+}): Promise<string> {
+  const fileLocation = fileAttachmentLocation({
+    workspaceId,
+    conversationId,
+    messageId,
+    contentFormat: "raw",
+  });
+
+  return getPrivateUploadBucket().getSignedUrl(fileLocation.filePath);
+}
+
+export async function renderContentFragmentForModel(
+  message: ContentFragmentType,
+  conversation: ConversationType,
+  model: ModelConfigurationType,
+  {
+    excludeImages,
+  }: {
+    excludeImages: boolean;
+  }
+): Promise<Result<ContentFragmentMessageTypeModel, Error>> {
+  const { contentType, sId, title } = message;
+
+  try {
+    if (isSupportedImageContentFragmentType(contentType)) {
+      if (excludeImages || !model.supportsVision) {
+        return new Ok({
+          role: "content_fragment",
+          name: `inject_${contentType}`,
+          content: [
+            {
+              type: "text",
+              text: `<attachment type="${contentType} title="${title}">[Image content interpreted by a vision-enabled model. Detailed description not available in this context.]`,
+            },
+          ],
+        });
+      }
+
+      const signedUrl = await getSignedUrlForRawContentFragment({
+        workspaceId: conversation.owner.sId,
+        conversationId: conversation.sId,
+        messageId: sId,
+      });
+
+      return new Ok({
+        role: "content_fragment",
+        name: `inject_${contentType}`,
+        content: [
+          {
+            type: "image_url",
+            image_url: {
+              url: signedUrl,
+            },
+          },
+        ],
+      });
+    } else {
+      const content = await getContentFragmentText({
+        workspaceId: conversation.owner.sId,
+        conversationId: conversation.sId,
+        messageId: sId,
+      });
+
+      return new Ok({
+        role: "content_fragment",
+        name: `inject_${contentType}`,
+        // The closing </attachment> tag will be added in the merging loop because we might
+        // need to add a "truncated..." mention in the selection loop.
+        content: [
+          {
+            type: "text",
+            text: `<attachment type="${contentType}" title="${title}">\n${content}\n`,
```

Comment:
Not super great that this logic gets separated from the main function given that this string does not include the closing tag yet.

I would suggest fully formatting the fragment here (which will simplify token count) and then do the right thing later in the main function when we truncate it 🙏

Reply:
Agree, but at the same time, all rendering logic should live in each message type. I can move here the token count so we properly format the content fragment directly.